hoomd/md/force.py

```diff
@@ -117,212 +118,144 @@ def additional_virial(self):
+    @property
+    def cpu_local_force_arrays(self):
+        """hoomd.data.LocalForceAccess: Expose force arrays on the CPU.
 
-        const.set_callback(updateForces)
-        run(100)
+        Provides direct access to the force, potential energy, torque, and
+        virial data of the particles in the system on the cpu through a context
+        manager. All data is MPI rank-local.
 
-        # Reset the callback
-        const.set_callback(None)
-    """
+        The `hoomd.data.LocalForceAccess` object returned by this property has
+        four arrays through which one can modify the force data:
+
+        Examples::
 
-    def set_callback(self, callback=None):  # noqa - will be ported to v3
-        self.cppForce.setCallback(callback)
+            with self.cpu_local_force_arrays as arrays:
+                arrays.force = ...
+                arrays.potential_energy = ...
+                arrays.torque = ...
+                arrays.virial = ...
+        """
+        if self._in_context_manager:
+            raise RuntimeError("Cannot enter cpu_local_force_arrays context "
+                               "manager inside another local_force_arrays "
+                               "context manager")
+        return hoomd.data.ForceLocalAccess(self)
+
+    @property
+    def gpu_local_force_arrays(self):
+        """hoomd.data.LocalForceAccessGPU: Expose force arrays on the GPU.
+
+        Provides direct access to the force, potential energy, torque, and
+        virial data of the particles in the system on the gpu through a context
+        manager. All data is MPI rank-local.
+
+        The `hoomd.data.LocalForceAccessGPU` object returned by this property
+        has four arrays through which one can modify the force data:
+
+        Examples::
+
+            with self.gpu_local_force_arrays as arrays:
+                arrays.force = ...
+                arrays.potential_energy = ...
+                arrays.torque = ...
+                arrays.virial = ...
+
+        Note:
+            GPU local force data is not available if the chosen device for the
+            simulation is `hoomd.device.CPU`.
+        """
+        if not isinstance(self._simulation.device, hoomd.device.GPU):
+            raise RuntimeError(
+                "Cannot access gpu_local_force_arrays without a GPU device")
+        if self._in_context_manager:
+            raise RuntimeError(
+                "Cannot enter gpu_local_force_arrays context manager inside "
+                "another local_force_arrays context manager")
+        return hoomd.data.ForceLocalAccessGPU(self)
+
+    @abstractmethod
+    def set_forces(self, timestep):
+        """Set the forces in the simulation loop.
 
-    # there are no coeffs to update in the constant force compute
-    def update_coeffs(self):  # noqa - will be ported to v3
+        Args:
+            timestep (int): The current timestep in the simulation.
```

Comment:
This seems like an odd parameter for something that is explicitly set or triggered. Might be helpful to clarify how it's used... which on my own looking up is:
To allow for easy access of the timestep for custom forces which are temporal in nature. Is this a significant enough usecase that we would discourage the perhaps uglier way of the users custom force class keeping a reference to the simulation object and getting it from there?

This is also making me think about efficiency if the force is constant. I'm just wanting to make sure I'm thinking about it correctly. Because the access is O(N(1)) there wouldn't be any benefit from caching these values if they were constant (the simpler old code that we're replacing). It's just a matter of if that those arrays would be kept around internally vs in a python object?

I think in general this is maybe pointing out to me that this might need some documentation about why this isn't based on custom.Action.

Reply:
The reason the timestep argument is here is because the `compute` method of classes that are derived from the `Compute` class in cpp all take the timestep as an argument, so I just forwarded that argument to the python method.

I don't think there is any reason to be more restrictive than what currently exists in the cpp side of the code. If a user doesn't want to use the timestep argument, they don't have to (in fact almost all computes operate this way).

One of the reasons behind adding the ability to write custom forces in python was so that people could prototype new potentials easily, and since we don't currently have any time-dependent potentials in cpp it seems reasonable to me that someone might want to prototype a time-dependent potential and see what happens.

The reason this isn't just a custom action is because the custom action API doesn't allow a user to change the forces on a set of particles. Using custom actions, you can access the system state via local snapshots, but you can't write your own force compute. That's where custom forces can help.